compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -0,0 +1,525 @@
+  , extRecEnvLvl1
+  , lookupVar, lookupRec, lookupVal, lookupChoice, lookupDataCon, lookupCid
+  , concatEnv
+  , emptyUpdateSet
+  , concatUpdateSet
+  , solveValueUpdatesEnv
+  , lookupChoInHMap
+  , fieldName2VarName
+  , recTypConFields, recTypFields, recExpFields
+  ) where
+
+import Control.Monad.Error.Class (MonadError (..), throwError)
+import Control.Monad.State.Lazy
+import Data.Hashable
+import GHC.Generics
+import Data.Maybe (listToMaybe, isJust)
+import Data.List (find)
+import Data.Bifunctor
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Text as T
+
+import DA.Daml.LF.Ast hiding (lookupChoice)
+import DA.Daml.LF.Verify.Subst
+
+-- | Data type denoting a create update.
+data UpdCreate = UpdCreate
+  { _creTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _creField :: ![(FieldName, Expr)]
+    -- ^ The fields to be verified, together with their value.
+  }
+  deriving Show
+-- | Data type denoting an archive update.
+data UpdArchive = UpdArchive
+  { _arcTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _arcField :: ![(FieldName, Expr)]
+    -- ^ The fields to be verified, together with their value.
+  }
+  deriving Show
+-- | Data type denoting an exercised choice update.
+data UpdChoice = UpdChoice
+  { _choTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _choName  :: !ChoiceName
+    -- ^ The name of the choice.
+  }
+  deriving Show
+
+-- | The List of updates being performed
+data UpdateSet = UpdateSet
+  { _usCre :: ![UpdCreate]
+    -- ^ The list of create updates.
+  , _usArc :: ![UpdArchive]
+    -- ^ The list of archive updates.
+  , _usCho :: ![UpdChoice]
+    -- ^ The list of choice updates.
+  , _usVal :: ![Qualified ExprValName]
+    -- ^ The list of referenced values. These will be replaced by their
+    -- respective updates after solving.
+    -- Note that this should be empty after the `ValuePhase`.
+  }
+  deriving Show
+
+-- | Create an empty update set.
+emptyUpdateSet :: UpdateSet
+emptyUpdateSet = UpdateSet [] [] [] []
+
+-- | Combine two update sets.
+concatUpdateSet :: UpdateSet
+  -- ^ The first update set to be combined.
+  -> UpdateSet
+  -- ^ The second update set to be combined.
+  -> UpdateSet
+concatUpdateSet (UpdateSet cres1 arcs1 chos1 vals1) (UpdateSet cres2 arcs2 chos2 vals2) =
+  UpdateSet (cres1 ++ cres2) (arcs1 ++ arcs2) (chos1 ++ chos2) (vals1 ++ vals2)
+
+-- | Refresh a given expression variable by producing a fresh renamed variable.
+-- TODO: when a renamed var gets renamed again, it might overlap again.
+-- We should have an additional field in VarName to denote it's number.
+genRenamedVar :: MonadEnv m
+  => ExprVarName
+  -- ^ The variable to be renamed.
+  -> m ExprVarName
+genRenamedVar (ExprVarName x) = ExprVarName . T.append x . T.pack <$> fresh
```

Comment:
This does not necessarily rename to a fresh name. For instance, `x` could be renamed to `x1` which is already in use. Is this intended?